basis/ui/tools/button-list/button-list.factor

```diff
@@ -0,0 +1,56 @@
+! Copyright (C) 2021 Kevin Cope.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors arrays assocs classes kernel math.rectangles 
+models models.arrow namespaces sequences ui ui.gadgets
+ui.gadgets.buttons ui.gadgets.glass ui.gadgets.labeled 
+ui.gadgets.presentations ui.gadgets.search-tables
+ui.gadgets.labels ui.gadgets.tables ui.gadgets.wrappers ui.gestures
+ui.theme ui.tools.browser.popups ui.tools.common ;
+FROM: ui.gadgets.wrappers => wrapper ;
+IN: ui.tools.button-list
+
+TUPLE: button-list-popup < wrapper ;
+
+MIXIN: clickable
+INSTANCE: button clickable
+
+SYMBOL: active-buttons
+active-buttons [ H{ } ] initialize
+
+: label-from-button ( button -- str/f )
+    children>> [ label? ] find swap [ text>> ] [ drop f ] if ;
+
+: store-labelled-button ( button -- str/f )
+    dup label-from-button [ [ active-buttons get set-at ] keep ] [ drop f ] if* ;
+
+: remove-labelled-button ( button -- str/f )
+    label-from-button [ dup active-buttons get delete-at ] [ f ] if* ;
+
+M: clickable graft* [ store-labelled-button drop ] [ call-next-method ] bi ;
+M: button ungraft* [ remove-labelled-button drop ] [ call-next-method ] bi ;
```

Comment:
Why is graft defined on ``clickable`` but ingraft on ``button``?

Reply:
Is fixed on the next commit, just a massive blunder on my part - static typing systems have made me complacent about these things